opensafely/log_stats.py

```diff
@@ -0,0 +1,156 @@
+        help="Project name (default: project-dir)",
+        type=str,
+    )
+    parser.add_argument(
+        "--output-file",
+        help="Output json filename (default: log_stats.json)",
+        type=Path,
+        default="log_stats.json",
+    )
+    return parser
+
+
+def _timestamp_for_honeytail(timestamp, ts_format):
+    honeytail_ts_format = "%Y-%m-%dT%H:%M:%S.000000000Z"
+    dt = datetime.strptime(timestamp, ts_format)
+    return dt.strftime(honeytail_ts_format)
+
+
+def parse_log(project_name, current_action, job_id, current_log):
+    current_log_timestamp = re.match(TIMESTAMP_PREFIX_REGEX, current_log).group()
+    current_log = re.sub(r"\s*\n\s*", " ", current_log).strip()
+    log_params = dict(re.findall(KEY_VALUE_REGEX, current_log))
+    ts_format = "%Y-%m-%dT%H:%M:%S"
+    return {
+        "timestamp": current_log_timestamp,
+        "project": project_name,
+        "job_id": job_id,
+        "action": current_action,
+        **log_params,
+    }
+
+
+def parse_stats_logs(raw_logs, project_name, current_action, job_id):
+    for log in raw_logs:
+        yield parse_log(project_name, current_action, job_id, log)
+
+
+def format_summary_stats(project_name, current_action, summary_stats):
+    start_time = summary_stats.get("created_at")
+    end_time = summary_stats.get("completed_at")
+    ts_format = "%Y-%m-%dT%H:%M:%SZ"
+    start_dt = datetime.strptime(start_time, ts_format)
+
+    if start_time and end_time:
+        elapsed = datetime.strptime(end_time, ts_format) - start_dt
+        summary_stats["elapsed_time_secs"] = elapsed.seconds
+        summary_stats["elapsed_time"] = str(elapsed)
+    return {
+        "timestamp": _timestamp_for_honeytail(start_time, ts_format),
+        "project": project_name,
+        "action": current_action,
+        **{k: v for k, v in summary_stats.items()},
+    }
+
+
+def main(project_dir, output_file, project_name=None):
+    log_dir = project_dir / "metadata"
+    project_name = project_name or project_dir.resolve().name
+
+    log_files = list((log_dir).glob("*.log"))
+
+    # Find the number of actions (equivalent to the number of log files)
+    action_count = len(log_files)
+    stats_logs = []
+    summary_stats_logs = []
+
+    for filep in log_files:
+        action_count += 1
+        current_action = filep.stem
+        # Include the total number of actions run alongside this one in each action summary
+        summary_stats = {"total_actions_in_job_request": action_count}
+        raw_logs = []
+        current_log = None
+        for line in filep.open().readlines():
+            # Logs in the log file can span multiple lines;
+            # check if this line is the beginning of a new log or the action summary
+            if re.match(TIMESTAMP_PREFIX_REGEX, line) or re.match(
+                COMPLETED_PREFIX_REGEX, line
+            ):
+                # If there's a current log, we're finished with it now, parse it add to the stats dict
```

Comment:
I don't understand this comment, I'm afraid.

Reply:
`current_log` is probably a bad name
Logs can span multiple lines; `current_log` is used to keep track of the one that's currently being parsed.  If line we're processing is the beginning of a new timestamped log, or the beginning of the "Completed" section of the log file, we must be done with the previous log, so we add it to the list of stats logs and reset `current_log` to this line.  (Also the comment is wrong, it used to parse it here too, but now it leaves it till the end and parses them all at once when it knows the job id and can add that in too)

Reply:
Does `current_log` contain the current line or the last line?